### PR TITLE
GraphQL-60: Manage Stored Payment Methods

### PR DESCRIPTION
**Pull request by pfantini**
### Description
https://github.com/magento/graphql-ce/issues/60
1. Create query to list customer payment tokens.
2. Create mutation to add/delete payment token.
3. Create functional test.

### Manual testing scenarios

1. Run the following request in GraphQL to get customer payment token list:
```
query {
  paymentTokens {
    public_hash
    payment_method_code
    type
    created_at
    expires_at
    details {
      code
      value
    }
  }
}
```

2. Run the following request in GraphQL to add new payment token:
```
mutation {
  paymentTokenAdd(input: {
    public_hash : "5555555"
    payment_method_code: "braintree"
    type: card
    expires_at: "2020-01-01 23:59:59"
    gateway_token: "ABC-123"
    details: [
          {
            code: "type",
            value: "DI"
          },
          {
            code: "maskedCC",
            value: "9876"
          },
          {
            code: "expirationDate",
            value: "12/2020"
          }
        ]
  }) {
    public_hash
    payment_method_code
    type
    created_at
    expires_at
    details {
      code
      value
    }
  }
}
```

3. Run the following request in GraphQL to delete payment token:
```
mutation {
  paymentTokenDelete(public_hash: "5555555")
}
```

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+type Mutation {
+    paymentTokenAdd(input: PaymentTokenInput!): PaymentToken @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenAdd") @doc(description: "Add payment tokens")
+    paymentTokenDelete(id: Int!): Boolean @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenDelete") @doc(description: "Remove stored payment tokens")
+}
+
+type PaymentToken @doc(description: "PaymentToken defines the stored payment details") {
+    entity_id: Int @doc(description: "The ID assigned to the payment token")
+    customer_id: Int @doc(description: "The customer ID")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]: No need to return `cutsomer_id` because customer can only query their own tokens

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    paymentTokenList: [PaymentToken] @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenList") @doc(description: "The paymentTokenList query returns a list of stored payment methods for a logged-in customer")
+}
+
+type Mutation {
+    paymentTokenAdd(input: PaymentTokenInput!): PaymentToken @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenAdd") @doc(description: "Add payment tokens")
+    paymentTokenDelete(id: Int!): Boolean @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenDelete") @doc(description: "Remove stored payment tokens")
+}
+
+type PaymentToken @doc(description: "PaymentToken defines the stored payment details") {
+    entity_id: Int @doc(description: "The ID assigned to the payment token")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    id: Int! @doc(description: "The ID assigned to the payment token")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    customer_id: Int @doc(description: "The customer ID")
+    public_hash: String @doc(description: "The public hash")
+    payment_method_code: String @doc(description: "The payment method code")
+    type: TokenType @doc(description: "The token type")
+    created_at: String @doc(description: "Timestamp indicating when the payment token was created")
+    expires_at: String @doc(description: "Timestamp indicating when the payment token was updated")
+    gateway_token: String @doc(description: "The getaway payment token")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]: It may be unsafe to expose this token, @joni-jones please confirm.
Same question regarding usage of the token in the input below.

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    details: [DetailsAttribute] @doc(description: "Payment token details")
+    is_active: Boolean @doc(description: "Indicating is active")
+    is_visible: Boolean @doc(description: "Indicating is visible")
+}
+
+type DetailsAttribute {
+    attribute_code: String @doc(description: "Attribute code")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    code: String! @doc(description: "Attribute code")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    value: String @doc(description: "Attribute value")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    value: String! @doc(description: "Attribute value")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+type Mutation {
+    paymentTokenAdd(input: PaymentTokenInput!): PaymentToken @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenAdd") @doc(description: "Add payment tokens")
+    paymentTokenDelete(id: Int!): Boolean @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenDelete") @doc(description: "Remove stored payment tokens")
+}
+
+type PaymentToken @doc(description: "PaymentToken defines the stored payment details") {
+    entity_id: Int @doc(description: "The ID assigned to the payment token")
+    customer_id: Int @doc(description: "The customer ID")
+    public_hash: String @doc(description: "The public hash")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    public_hash: String! @doc(description: "The public hash")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    payment_method_code: String @doc(description: "The payment method code")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    payment_method_code: String! @doc(description: "The payment method code")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    type: TokenType @doc(description: "The token type")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    type: TokenType! @doc(description: "The token type")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    created_at: String @doc(description: "Timestamp indicating when the payment token was created")
+    expires_at: String @doc(description: "Timestamp indicating when the payment token was updated")
+    gateway_token: String @doc(description: "The getaway payment token")
+    details: [DetailsAttribute] @doc(description: "Payment token details")
+    is_active: Boolean @doc(description: "Indicating is active")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    is_active: Boolean! @doc(description: "Indicating is active")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    is_visible: Boolean @doc(description: "Indicating is visible")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    is_visible: Boolean! @doc(description: "Indicating is visible")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    paymentTokenDelete(id: Int!): Boolean @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenDelete") @doc(description: "Remove stored payment tokens")
+}
+
+type PaymentToken @doc(description: "PaymentToken defines the stored payment details") {
+    entity_id: Int @doc(description: "The ID assigned to the payment token")
+    customer_id: Int @doc(description: "The customer ID")
+    public_hash: String @doc(description: "The public hash")
+    payment_method_code: String @doc(description: "The payment method code")
+    type: TokenType @doc(description: "The token type")
+    created_at: String @doc(description: "Timestamp indicating when the payment token was created")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    created_at: String! @doc(description: "Timestamp indicating when the payment token was created")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    expires_at: String @doc(description: "Timestamp indicating when the payment token was updated")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    expires_at: String! @doc(description: "Timestamp indicating when the payment token was updated")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls (hunk 7 above)
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    payment_method_code: String! @doc(description: "The payment method code")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls (hunk 3 above)
Comment (YevSent) [reply app/code/Magento/VaultGraphQl/etc/schema.graphqls]: @paliarus, you are right. The `gateway_token` should not be exposed. Only `public_hash` should be available on the storefront.

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    type: TokenType! @doc(description: "The payment token type")
+    expires_at: String @doc(description: "Timestamp indicating when the payment token was updated")
+    gateway_token: String! @doc(description: "The getaway payment token")
+    details: [DetailsAttributeInput] @doc(description: "Payment token details")
+    is_active: Boolean @doc(description: "Indicating is active")
+    is_visible: Boolean @doc(description: "Indicating is visible")
+}
+
+input DetailsAttributeInput {
+    attribute_code: String @doc(description: "Attribute code")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    attribute_code: String! @doc(description: "Attribute code")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    value: String @doc(description: "Attribute value")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    value: String! @doc(description: "Attribute value")
```

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls (hunk 2 above)
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]: Do we need to expose the ID? Can the client work with public_hash in all operations instead?

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 15]
```diff
@@ -0,0 +1,51 @@
+# Copyright © Magento, Inc. All rights reserved.
+# See COPYING.txt for license details.
+
+type Query {
+    paymentTokenList: [PaymentToken] @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenList") @doc(description: "The paymentTokenList query returns a list of stored payment methods for a logged-in customer")
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]:
```suggestion
    paymentTokens: [PaymentToken] @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenList") @doc(description: "The paymentTokenList query returns a list of stored payment methods for a logged-in customer")
```

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        'public_hash',
+        'payment_method_code',
+        'type',
+        'gateway_token'
+    ];
+
+    /**
+     * @var PaymentTokenRepositoryInterface
+     */
+    private $paymentTokenRepositoryInterface;
```
Comment (YevSent) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php]: Redundant `interface` word.

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+                return $attributeName;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Process payment token add
+     *
+     * @param $customerId
```
Comment (YevSent) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php]: An argument without type.

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+     * @param array $tokenInfo
+     * @return PaymentTokenInterface
+     * @throws GraphQlInputException
+     * @throws GraphQlAlreadyExistsException
+     */
+    private function processPaymentTokenAdd($customerId, array $tokenInfo) : PaymentTokenInterface
```
Comment (YevSent) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php]: Redundant space before `:`

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+        }
+        /** @var PaymentTokenInterface $token */
+        $token = $this->paymentTokenDataProvider->fillPaymentToken(
+            $this->paymentTokenFactoryInterface->create($tokenInfo['type']),
+            $tokenInfo
+        );
+        $token->setCustomerId($customerId);
+        try {
+            return $this->paymentTokenRepositoryInterface->save($token);
+        }catch (AlreadyExistsException $e) {
```
Comment (YevSent) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php]: Missed space between `}` and `catch`

**Review by YevSent — CHANGES_REQUESTED** [review]

**Comment by YevSent** [comment]
The current implementation of Checkout storefront doesn't use token id, the public hash is used instead. I would recommend avoiding the usage of entity id field.

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentToken/PaymentTokenDataProvider.php [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+        $result = [];
+        /** @var PaymentTokenInterface $paymentToken */
+        foreach ($paymentTokens as $paymentToken) {
+            $result[] = $this->processPaymentToken($paymentToken);
+        }
+        return $result;
+    }
+
+    /**
+     * Transform single payment token data from object to in array format
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentToken/PaymentTokenDataProvider.php]:
```suggestion
     * Transform single payment token data from object to an array format
```
Same in other comments

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentToken/PaymentTokenDataProvider.php [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+        $this->dataObjectHelper = $dataObjectHelper;
+    }
+
+    /**
+     * Transform array of payment token data from object to in array format
+     *
+     * @param PaymentTokenInterface[] $paymentTokens
+     * @return array
+     */
+    public function processPaymentTokens(array $paymentTokens) : array
```
Comment (paliarush) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentToken/PaymentTokenDataProvider.php]: The method name is non-descriptive. Please change to a more meaningful one

File: dev/tests/integration/testsuite/Magento/Vault/_files/payment_active_tokens.php [hunk 22]
```diff
@@ -0,0 +1,62 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+include "customer.php";
```
Comment (paliarush) [inline dev/tests/integration/testsuite/Magento/Vault/_files/payment_active_tokens.php:7]: Please implement `rollback` script to remove test data persisted by this fixture.

**Review by paliarush — CHANGES_REQUESTED** [review]
@pfantini thanks for the contribution, please address review comments above.

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+        $context,
+        ResolveInfo $info,
+        array $value = null,
+        array $args = null
+    ) {
+        /** @var ContextInterface $context */
+        if ((!$context->getUserId()) || $context->getUserType() == UserContextInterface::USER_TYPE_GUEST) {
+            throw new GraphQlAuthorizationException(
+                __(
+                    'Current customer does not have access to the resource "%1"',
```
Comment (keharper) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php]: Change exception message to

A guest customer cannot access resource "%1".

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+     * @return PaymentTokenInterface
+     * @throws GraphQlInputException
+     * @throws GraphQlAlreadyExistsException
+     */
+    private function processPaymentTokenAdd($customerId, array $tokenInfo) : PaymentTokenInterface
+    {
+        $errorInput = $this->getInputError($tokenInfo);
+        if ($errorInput) {
+            throw new GraphQlInputException(
+                __('Required parameter %1 is missing', [$errorInput])
```
Comment (keharper) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenAdd.php]: Change message to 

The required parameter %1 is missing.

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenDelete.php [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+        $context,
+        ResolveInfo $info,
+        array $value = null,
+        array $args = null
+    ) {
+        /** @var ContextInterface $context */
+        if ((!$context->getUserId()) || $context->getUserType() == UserContextInterface::USER_TYPE_GUEST) {
+            throw new GraphQlAuthorizationException(
+                __(
+                    'Current customer does not have access to the resource "%1"',
```
Comment (keharper) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenDelete.php]: Change exception message to

A guest customer cannot access resource "%1".

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenDelete.php [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+     * @throws GraphQlAuthorizationException
+     * @throws GraphQlNoSuchEntityException
+     */
+    private function deleteToken($customerId, $tokenId)
+    {
+        /** @var PaymentTokenInterface $token */
+        $token = $this->paymentTokenRepositoryInterface->getById($tokenId);
+        if (empty($token->getEntityId())) {
+            throw new GraphQlNoSuchEntityException(
+                __('Payment Token id %1 does not exist.', [$tokenId])
```
Comment (keharper) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenDelete.php]: Change exception message to

Payment token ID %1 does not exist.

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenDelete.php [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+            );
+        }
+        if ($customerId != $token->getCustomerId()) {
+            throw new GraphQlAuthorizationException(
+                __('Current customer does not have permission to delete Payment Token id %1', [$tokenId])
```
Comment (keharper) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenDelete.php]: If this exception applies to guests only, then

A guest customer cannot delete a payment token.

If this applies to a logged-in customer, then

The current customer does not have permission to delete payment token ID %1.

Actually, I would like to see the customer ID in the exception message.

File: app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenList.php [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+        $context,
+        ResolveInfo $info,
+        array $value = null,
+        array $args = null
+    ) {
+        /** @var ContextInterface $context */
+        if ((!$context->getUserId()) || $context->getUserType() == UserContextInterface::USER_TYPE_GUEST) {
+            throw new GraphQlAuthorizationException(
+                __(
+                    'Current customer does not have access to the resource "%1"',
```
Comment (keharper) [inline app/code/Magento/VaultGraphQl/Model/Resolver/PaymentTokenList.php]: Change exception message to

A guest customer cannot access resource "%1".

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+type Mutation {
+    paymentTokenAdd(input: PaymentTokenInput!): PaymentToken @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenAdd") @doc(description: "Add payment tokens")
+    paymentTokenDelete(id: Int!): Boolean @resolver(class: "Magento\\VaultGraphQl\\Model\\Resolver\\PaymentTokenDelete") @doc(description: "Remove stored payment tokens")
+}
+
+type PaymentToken @doc(description: "PaymentToken defines the stored payment details") {
```
Comment (keharper) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]: Change the description to 

PaymentToken defines stored payment details

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls (hunk 9 above)
Comment (keharper) [reply app/code/Magento/VaultGraphQl/etc/schema.graphqls]: Change desciption to 

Indicates the [payment token? field? Insert noun here] is active

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls (hunk 10 above)
Comment (keharper) [reply app/code/Magento/VaultGraphQl/etc/schema.graphqls]: Change desciption to 

Indicates the [payment token? field? Insert noun here] is visible

File: app/code/Magento/VaultGraphQl/etc/schema.graphqls [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+input PaymentTokenInput {
+    public_hash: String! @doc(description: "The public hash")
+    payment_method_code: String! @doc(description: "The payment method code")
+    type: TokenType! @doc(description: "The payment token type")
+    expires_at: String @doc(description: "Timestamp indicating when the payment token was updated")
+    gateway_token: String! @doc(description: "The getaway payment token")
+    details: [DetailsAttributeInput] @doc(description: "Payment token details")
+    is_active: Boolean @doc(description: "Indicating is active")
```
Comment (keharper) [inline app/code/Magento/VaultGraphQl/etc/schema.graphqls]: Change the description for is_active and is_visible to be the same as above.

**Comment by YevSent** [comment]
I don't see the point to create GraphQL request to add a new token, most of the payment integrations allow storing the token after authorization transaction and the token might be received only in a server-server call.

**Review by paliarush — APPROVED** [review]
@pfantini All changes after the initial CR look good, thank you.
Please address @joni-jones 's concern and we can merge this PR.

File: lib/internal/Magento/Framework/Setup/Declaration/Schema/Declaration/SchemaBuilder.php [hunk 31]
```diff
@@ -129,7 +129,7 @@ private function validate(Schema $schema)
                 $messages .= sprintf("%s%s", PHP_EOL, $error['message']);
             }
 
-            throw new Exception(new Phrase($messages));
+            //throw new Exception(new Phrase($messages));
```
Comment (naydav) [inline lib/internal/Magento/Framework/Setup/Declaration/Schema/Declaration/SchemaBuilder.php]: Need to revert this change

**Comment by naydav** [comment]
Functionality was merged with https://github.com/magento/graphql-ce/pull/330 and https://github.com/magento/graphql-ce/pull/307
